go-controller/pkg/node/gateway_localnet.go

```diff
@@ -179,45 +156,76 @@ func initLocalnetGateway(nodeName string, subnet string, wf *factory.WatchFactor
 		return err
 	}
 
+	var gatewayIP, gatewayNextHop, gatewaySubnetPrefix string
+	if utilnet.IsIPv6CIDRString(subnet) {
+		gatewayIP = v6localnetGatewayIP
+		gatewayNextHop = v6localnetGatewayNextHop
+		gatewaySubnetPrefix = v6localnetGatewaySubnetPrefix
+	} else {
+		gatewayIP = v4localnetGatewayIP
+		gatewayNextHop = v4localnetGatewayNextHop
+		gatewaySubnetPrefix = v4localnetGatewaySubnetPrefix
+	}
+
 	// Flush any addresses on localnetBridgeNextHopPort and add the new IP address.
 	if err = util.LinkAddrFlush(link); err == nil {
-		err = util.LinkAddrAdd(link, localnetGatewayNextHopSubnet())
+		err = util.LinkAddrAdd(link, gatewayNextHop+gatewaySubnetPrefix)
 	}
 	if err != nil {
 		return err
 	}
 
 	err = util.SetLocalL3GatewayConfig(nodeAnnotator, ifaceID, macAddress,
-		localnetGatewayIP(), localnetGatewayNextHop(),
+		gatewayIP+gatewaySubnetPrefix, gatewayNextHop,
 		config.Gateway.NodeportEnable)
 	if err != nil {
 		return err
 	}
 
-	if config.IPv6Mode {
+	if utilnet.IsIPv6CIDRString(subnet) {
 		// TODO - IPv6 hack ... for some reason neighbor discovery isn't working here, so hard code a
 		// MAC binding for the gateway IP address for now - need to debug this further
-		err = util.LinkNeighAdd(link, "fd99::2", macAddress)
+		err = util.LinkNeighAdd(link, gatewayIP, macAddress)
 		if err == nil {
-			klog.Infof("Added MAC binding for fd99::2 on %s", localnetGatewayNextHopPort)
+			klog.Infof("Added MAC binding for %s on %s", gatewayIP, localnetGatewayNextHopPort)
 		} else {
-			klog.Errorf("Error in adding MAC binding for fd99::2 on %s: %v", localnetGatewayNextHopPort, err)
+			klog.Errorf("Error in adding MAC binding for %s on %s: %v", gatewayIP, localnetGatewayNextHopPort, err)
 		}
 	}
 
-	err = localnetGatewayNAT(ipt, localnetGatewayNextHopPort, localnetGatewayIP())
+	ipt, err := localnetIPTablesHelper(subnet)
+	if err != nil {
+		return err
+	}
+
+	err = localnetGatewayNAT(ipt, localnetGatewayNextHopPort, gatewayIP)
 	if err != nil {
 		return fmt.Errorf("Failed to add NAT rules for localnet gateway (%v)", err)
 	}
 
 	if config.Gateway.NodeportEnable {
-		err = localnetNodePortWatcher(ipt, wf)
+		err = localnetNodePortWatcher(ipt, wf, gatewayIP)
 	}
 
 	return err
 }
 
-func localnetIptRules(svc *kapi.Service) []iptRule {
+// localnetIPTablesHelper gets an IPTablesHelper for IPv4 or IPv6 as appropriate
+func localnetIPTablesHelper(subnet string) (util.IPTablesHelper, error) {
```

Comment:
@danwinship : what does this mean for us in a dual-stack world with services which can be both IPv6 and IPv4? I guess this is incorrect in such a case? We should not define a global helper, but instead instantiate one for each service as to ascertain the correct IP family and act accordingly (which is the way shared-gateway mode does it: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/node/gateway_shared_intf_linux.go#L167).

I can take care of this in my ExternalIP PR, but I just need your confirmation that my understanding is correct.

Reply:
Or be smarter than that and define a map of iptables-helpers for each family and just re-use that for every service we process.